File: sdks/python/ags_sdk/keystore.py
"""Local keystore, interoperable with the ``ags`` CLI."""

import json
import os
import stat

from .errors import AGS1Error
from . import keys as keys_mod


def default_path() -> str:
    return os.path.join(os.path.expanduser("~"), ".ags", "credentials.json")


def load_keystore(path=None) -> dict:
    """Load a keystore written by the ``ags`` CLI or by save_keystore.

    Refuses a file readable by other users. A key already exposed to other
    local accounts should be rotated, not used, so this is an error rather than
    a warning.
    """
    path = path or default_path()

    if not os.path.exists(path):
        raise AGS1Error(f"no keystore at {path}")

    mode = stat.S_IMODE(os.stat(path).st_mode)
    if mode & 0o077:
        raise AGS1Error(
            f"{path} is readable by other users (mode {mode:04o}); "
            f"run 'chmod 600 {path}' and rotate this key, since it may already be compromised"
        )

    with open(path, "r", encoding="utf-8") as handle:
        record = json.load(handle)

    private_key = keys_mod.load_private_key(record["privateKey"])

    # The stored kid must still match the stored key material, so a hand-edited
    # file cannot point the SDK at a credential it does not hold the key for.
    derived = keys_mod.key_id(private_key.public_key())
    declared = record.get("keyId")
    if declared and declared != derived:
        raise AGS1Error(
            f"keystore key id {declared!r} does not match the stored key material ({derived!r})"
        )
    record["keyId"] = derived
    record["_private_key"] = private_key

    return record


def save_keystore(path, tenant_id, agent_id, private_key, gateway_url=None) -> str:
    """Write a keystore with 0600 permissions.

    The write goes to a temporary file in the same directory and is then
    renamed, so an interrupted write cannot leave a truncated key file behind.
    """
    import base64
    import tempfile
    from datetime import datetime, timezone

    path = path or default_path()
    directory = os.path.dirname(path) or "."
    os.makedirs(directory, mode=0o700, exist_ok=True)

    public_key = private_key.public_key()
    record = {
        "tenantId": tenant_id,
        "agentId": agent_id,
        "keyId": keys_mod.key_id(public_key),
        "privateKey": base64.b64encode(keys_mod.private_key_bytes(private_key)).decode("ascii"),
        "publicKey": base64.b64encode(keys_mod.public_key_bytes(public_key)).decode("ascii"),
        "createdAt": datetime.now(timezone.utc).isoformat(),
    }
    if gateway_url:
        record["gatewayUrl"] = gateway_url

    fd, tmp = tempfile.mkstemp(dir=directory, prefix=".credentials-", suffix=".tmp")
    try:
        os.fchmod(fd, 0o600)
        with os.fdopen(fd, "w", encoding="utf-8") as handle:
            json.dump(record, handle, indent=2)
            handle.flush()
            os.fsync(handle.fileno())
        os.replace(tmp, path)
    except Exception:
        if os.path.exists(tmp):
            os.unlink(tmp)
        raise

    return path

File: sdks/go/agentgw/keystore/keystore.go
// Package keystore persists AGS1 agent keys on local disk.
package keystore

import (
	"crypto/ed25519"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/Aryan22g/agw/pkg/ags1/keys"
)

// Record is one stored agent identity.
type Record struct {
	TenantID   string    `json:"tenantId"`
	AgentID    string    `json:"agentId"`
	KeyID      string    `json:"keyId"`
	PrivateKey string    `json:"privateKey"`
	PublicKey  string    `json:"publicKey"`
	CreatedAt  time.Time `json:"createdAt"`
	GatewayURL string    `json:"gatewayUrl,omitempty"`
}

// ErrNotFound is returned when no key file exists at the given path.
var ErrNotFound = errors.New("keystore: no key found")

// FileStore reads and writes a single key file.
//
// The file holds an unencrypted private key, so its permissions are the only
// thing protecting it. Load refuses to read a file that is group- or
// world-readable rather than warning: a key that has already been exposed to
// other local users should be rotated, not used.
type FileStore struct {
	path string
}

// NewFileStore builds a store over a path. An empty path resolves to
// ~/.ags/credentials.json.
func NewFileStore(path string) (*FileStore, error) {
	if strings.TrimSpace(path) == "" {
		home, err := os.UserHomeDir()
		if err != nil {
			return nil, fmt.Errorf("keystore: resolve home dir: %w", err)
		}
		path = filepath.Join(home, ".ags", "credentials.json")
	}
	return &FileStore{path: filepath.Clean(path)}, nil
}

// Path returns the file this store reads and writes.
func (s *FileStore) Path() string { return s.path }

// Save writes a record with 0600 permissions.
//
// The write goes to a temporary file in the same directory and is then
// renamed, so an interrupted write cannot leave a truncated key file behind.
func (s *FileStore) Save(rec Record) error {
	dir := filepath.Dir(s.path)
	if err := os.MkdirAll(dir, 0o700); err != nil {
		return fmt.Errorf("keystore: create dir: %w", err)
	}

	data, err := json.MarshalIndent(rec, "", "  ")
	if err != nil {
		return fmt.Errorf("keystore: encode record: %w", err)
	}

	tmp, err := os.CreateTemp(dir, ".credentials-*.tmp")
	if err != nil {
		return fmt.Errorf("keystore: create temp file: %w", err)
	}
	tmpName := tmp.Name()
	defer func() { _ = os.Remove(tmpName) }()

	if err := tmp.Chmod(0o600); err != nil {
		tmp.Close()
		return fmt.Errorf("keystore: chmod temp file: %w", err)
	}
	if _, err := tmp.Write(data); err != nil {
		tmp.Close()
		return fmt.Errorf("keystore: write temp file: %w", err)
	}
	if err := tmp.Sync(); err != nil {
		tmp.Close()
		return fmt.Errorf("keystore: sync temp file: %w", err)
	}
	if err := tmp.Close(); err != nil {
		return fmt.Errorf("keystore: close temp file: %w", err)
	}

	if err := os.Rename(tmpName, s.path); err != nil {
		return fmt.Errorf("keystore: install key file: %w", err)
	}
	return nil
}

// Load reads a record, refusing files with unsafe permissions.
func (s *FileStore) Load() (*Record, error) {
	info, err := os.Stat(s.path)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, fmt.Errorf("%w at %s", ErrNotFound, s.path)
		}
		return nil, fmt.Errorf("keystore: stat: %w", err)
	}

	if perm := info.Mode().Perm(); perm&0o077 != 0 {
		return nil, fmt.Errorf(
			"keystore: %s is readable by other users (mode %04o); "+
				"run 'chmod 600 %s' and rotate this key, since it may already be compromised",
			s.path, perm, s.path)
	}

	data, err := os.ReadFile(s.path)
	if err != nil {
		return nil, fmt.Errorf("keystore: read: %w", err)
	}

	var rec Record
	if err := json.Unmarshal(data, &rec); err != nil {
		return nil, fmt.Errorf("keystore: parse: %w", err)
	}

	// The stored kid must still match the stored key material, so a hand
	// edited file cannot point the SDK at a credential it does not hold the
	// key for.
	priv, err := keys.DecodePrivateKey(rec.PrivateKey)
	if err != nil {
		return nil, fmt.Errorf("keystore: %w", err)
	}
	derived, err := keys.KeyID(priv.Public().(ed25519.PublicKey))
	if err != nil {
		return nil, fmt.Errorf("keystore: derive key id: %w", err)
	}
	if rec.KeyID != "" && rec.KeyID != derived {
		return nil, fmt.Errorf("keystore: stored key id %q does not match the stored key material (%q)",
			rec.KeyID, derived)
	}
	rec.KeyID = derived

	return &rec, nil
}
